python/ray/ml/train/impl/function_trainer.py

```diff
@@ -0,0 +1,55 @@
+import abc
+from typing import Dict, Any, Optional
+
+from ray.ml.checkpoint import Checkpoint
+from ray.ml.train import Trainer
+from ray.tune import PlacementGroupFactory
+from ray.util.annotations import DeveloperAPI
+
+
+@DeveloperAPI
+class FunctionTrainer(Trainer):
+    """A Trainer that provides a function interface.
+
+    When subclassing ``FunctionTrainer``, you only need to override the
+    ``training_function`` and ``resource_function`` methods. There is no
+    need to directly work with Trainables.
+
+    Args:
+        Same as ray.ml.Trainer.
+
+    """
+
+    def __init__(self, *args, **kwargs):
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def training_function(
+        self, config: Dict[str, Any], checkpoint: Optional[Checkpoint] = None
+    ):
+        """Function to define the training logic.
+
+        Implement this the same way you would implement a training function
+        for Tune.
```

Comment:
Can we add a reference for this and an example? What is a "training function for Tune"? 

Also, long-term we should come up with a crystal clear naming convention for this. We sometimes call this a Trainable (function), a training function, or an objective (function).

Reply:
For this function signature, can we accept all relevant arguments directly? E.g. `training_function(datasets, config, checkpoint)` or so. There are two reasons for this, 1) datasets is central to the Trainer and users will always want to access the processed datasets (same with config and checkpoint), and 2) it allows us more flexibility, e.g. we could move to true "function trainers" more easily (no class definition, just runnable training function, like in ray tune and train)

Reply:
@maxpumperla I removed the sentence referencing Tune. I'm still not sure exactly how to describe this (it makes a lot of sense if you're already familiar with the Tune function API, but if not, then it probably doesn't). Let me think about this some more. Any suggestions are very welcome here.

Reply:
@krfricke these are great points, I'll surface our discussions on this to the broader group